src/hir/combiner.rs

```diff
@@ -0,0 +1,90 @@
+pub struct Combiner {}
+
+impl Combiner {
+    /// Creates a new combiner.
+    #[must_use]
+    pub fn new() -> Self {
+        Combiner {}
+    }
+
+    /// Splits the input text into distinct trees,
+    /// delimited by two successive newlines
+    ///
+    /// This function is called before the tokenization and parsing steps.
+    pub fn split<'a>(&'a self, text: &'a str) -> Vec<&str> {
+        text.split("\n\n").collect::<Vec<&str>>()
+    }
+
+    /// Verifies that the ASTs are valid.
+    ///
+    /// This function is called before the translation to HIR.
+    pub fn verify(&self, asts: &Vec<Ast>) -> anyhow::Result<(), String> {
+        if let Some(Ast::Root(first_root)) = asts.first() {
+            let first_contract_name = &first_root.contract_name;
+            let first_parts: Vec<&str> = first_contract_name
+                .split(|c| c == '.' || c == ':')
+                .collect();
+
+            for ast in asts {
+                if let Ast::Root(root) = ast {
+                    let parts: Vec<&str> =
+                        root.contract_name.split(|c| c == '.' || c == ':').collect();
+                    if parts[0] != first_parts[0] {
+                        return Err(format!(
+                            "Contract name mismatch: expected '{}', found '{}'",
+                            first_contract_name, root.contract_name
+                        ));
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Combines the translated HIRs into a single HIR. HIRs are merged by
+    /// iterating over each HIR and merge their children into the root,
+    /// while filtering out duplicate modifiers.
+    ///
+    /// This function is called after the ASTs are translated to HIR.
+    pub fn combine(&self, hirs: &[Hir]) -> Hir {
+        let mut root: Root = Root::default();
+        let mut added_modifiers = HashSet::new();
+
+        for hir in hirs {
+            match hir {
+                Hir::Root(r) => {
+                    // If child is of type FunctionDefinition with the same identifier as a child of another root of ty
+                    // Modifier, then they are duplicates. Traverse all children of the root and remove the duplicates.
+                    for child in &r.children {
+                        match child {
+                            Hir::FunctionDefinition(func_def)
+                                if func_def.ty == super::FunctionTy::Modifier =>
+                            {
+                                if added_modifiers.insert(func_def.identifier.clone()) {
+                                    root.children.push(child.clone());
+                                }
+                            }
+                            _ => root.children.push(child.clone()),
+                        }
+                    }
+                }
+                _ => unreachable!(),
+            }
+        }
+
+        Hir::Root(root)
+    }
```

Comment:
No advantage, I think I just originally split these steps when breaking down the problem into its constituent parts but agree that they should be performed together in a single function.

Reply:
Ah, actually, having looked at this again more closely, I split these steps i. so that we fail early if the contract names do not match, and ii. because it was easier to verify the `contract_name` member on `Ast::Root` rather than finding all the `Hir::ContractDefinition`s and comparing `identifier`s. In doing so, I just realized that we actually don't want to push all children that are `Hir::ContractDefinition`s to the `Root` but rather only one of these and the rest of the `Hir::FunctionDefinition`s pushed to this child, so I need to update this anyway and will modify the verification to work with identifiers instead.

Reply:
> I just realized that we actually don't want to push all children that are Hir::ContractDefinitions to the Root but rather only one of these and the rest of the Hir::FunctionDefinitions pushed to this child, so I need to update this anyway and will modify the verification to work with identifiers instead.

This makes sense to me 👍🏻 

> i. so that we fail early if the contract names do not match

Just want to comment here that I think we don't want to fail early here: As a user, I'd like to see all the contract name mismatches at once, instead of one at a time. It can be frustrating to fix the name in one tree only to run bulloak again and find out that there was another name mismatch.

This doesn't need to be implemented here, but it's just something to keep in mind.

Reply:
Great! And I meant fail early in the sense that we verify all contract names match before combining then but yeah realise that the current implementation throws mismatch errors one at a time so will fix that with the other error-handling suggestions elsewhere.